webrender/src/renderer.rs

```diff
@@ -514,107 +512,79 @@ impl CacheTexture {
     }
 }
 
-
-trait GpuStoreLayout {
-    fn image_format() -> ImageFormat;
-
-    fn texture_width<T>() -> usize;
-
-    fn texture_filter() -> TextureFilter;
-
-    fn texel_size() -> usize {
-        match Self::image_format() {
-            ImageFormat::BGRA8 => 4,
-            ImageFormat::RGBAF32 => 16,
-            _ => unreachable!(),
-        }
-    }
-
-    fn texels_per_item<T>() -> usize {
-        let item_size = mem::size_of::<T>();
-        let texel_size = Self::texel_size();
-        debug_assert!(item_size % texel_size == 0);
-        item_size / texel_size
-    }
-
-    fn items_per_row<T>() -> usize {
-        Self::texture_width::<T>() / Self::texels_per_item::<T>()
-    }
-
-    fn rows_per_item<T>() -> usize {
-        Self::texels_per_item::<T>() / Self::texture_width::<T>()
-    }
-}
-
-struct GpuDataTexture<L> {
+struct VertexDataTexture {
     id: TextureId,
-    layout: PhantomData<L>,
+    pbo: PBOId,
 }
 
-impl<L: GpuStoreLayout> GpuDataTexture<L> {
-    fn new(device: &mut Device) -> GpuDataTexture<L> {
+impl VertexDataTexture {
+    fn new(device: &mut Device) -> VertexDataTexture {
         let id = device.create_texture_ids(1, TextureTarget::Default)[0];
+        let pbo = device.create_pbo();
 
-        GpuDataTexture {
+        VertexDataTexture {
             id,
-            layout: PhantomData,
+            pbo,
         }
     }
 
-    fn init<T: Default>(&mut self,
-                        device: &mut Device,
-                        data: &mut Vec<T>) {
+    fn update<T>(&mut self,
+                 device: &mut Device,
+                 data: &mut Vec<T>) {
         if data.is_empty() {
             return;
         }
 
-        let items_per_row = L::items_per_row::<T>();
-        let rows_per_item = L::rows_per_item::<T>();
+        debug_assert!(mem::size_of::<T>() % 16 == 0);
+        let texels_per_item = mem::size_of::<T>() / 16;
+        let items_per_row = MAX_VERTEX_TEXTURE_WIDTH / texels_per_item;
 
         // Extend the data array to be a multiple of the row size.
         // This ensures memory safety when the array is passed to
         // OpenGL to upload to the GPU.
         if items_per_row != 0 {
             while data.len() % items_per_row != 0 {
-                data.push(T::default());
+                data.push(unsafe { mem::uninitialized() });
```

Comment:
I never understood the obsession with `uninitialized` as opposed to `zeroed`. It reads like a premature optimization to me, which may end up wasting debugging time when something empty would look like it has data.

Reply:
Fair point - will change to zeroed() as a follow up.